File: Day 7/Permutation in String.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        n,m = len(s1), len(s2)
        if n > m:
            return False

        need = Counter(s1)
        window = Counter(s2[:n])

        if window == need:
            return True

        for i in range(n, m):
            add_char = s2[i]
            rem_char = s2[i-n]

            window[add_char] +=1
            window[rem_char] -=1

            if window[rem_char] == 0:
                del window[rem_char]

            if window == need:
                return True
        return False

File: Day 7/Permutation in String.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int n = s1.size(), m = s2.size();
        if (n > m){
            return false;
        }
        vector<int> need(26, 0);
        vector<int> window(26, 0);
        // count frequencies of s1
        for (char c : s1) {
            need[c - 'a']++;
        }

        // first window on s2
        for (int i = 0; i < n; i++) {
            window[s2[i] - 'a']++;
        }
        if (window == need){
            return true;
        }

        for (int i =n; i < m; i++ ){
            window[s2[i]- 'a']++;
            window[s2[i - n]- 'a']--;
            if (window == need) return true;
        }

        return false;

    }
};
